spring-kafka/src/main/java/org/springframework/kafka/listener/RecordInterceptor.java

```diff
@@ -81,4 +81,29 @@ default void success(ConsumerRecord<K, V> record, Consumer<K, V> consumer) {
 	default void failure(ConsumerRecord<K, V> record, Exception exception, Consumer<K, V> consumer) {
 	}
 
+	/**
+	 * Called before consumer is polled.
+	 * <p>
+	 * It can be used to set up thread-bound resources which will be available for the
+	 * entire duration of the consumer poll operation e.g. logging with MDC.
+	 * </p>
+	 * @param consumer the consumer.
+	 * @since 2.8
+	 */
+	default void beforePoll(Consumer<K, V> consumer) {
+	}
+
+	/**
+	 * Called after records were processed by listener and error handler.
+	 * <p>
+	 * It can be used to clear thread-bound resources which were set up in {@link #beforePoll(Consumer)}.
+	 * This is the last method called by the {@link MessageListenerContainer} before the next record
+	 * processing cycle starts.
+	 * </p>
+	 * @param consumer the consumer.
+	 * @since 2.8
+	 */
+	default void afterRecordsProcessed(Consumer<K, V> consumer) {
```

Comment:
This one is for a single record interceptor, so the contract must be called `afterRecordProcessed` and have a `ConsumerRecord<K, V> record` as a first argument.
I might miss something and probably this per-record interceptor is not designed for an "after processed" hook...
Otherwise it sounds awkward to implement a `RecordInterceptor` and figure out somehow a contract for the whole batch as it sounds right now.

Reply:
No; this would be redundant - we already call success or failure for each record.

This hook is to clean up state when all records have been processed - the `beforePoll()` is intended so that thread-bound resources can be registered and made available in components called by the `KafkaConsumer` within the `poll()` (e.g. deserializers, partition assignors, etc, etc).

This hook is to unbind those resources.

Reply:
Right; so should not be there then some other contract not the one tied to the interceptors?
It feel right now for me that this is not an interception  logic: we just don't intercept anything before and after.
Feels more like a `BeforeAfterPollProcessor`.
No? What was a premise to put such a logic into the interceptor contract?

Reply:
It was [my suggestion on the issue](https://github.com/spring-projects/spring-kafka/issues/1704#issuecomment-896997190).

However, I see that we could.should extract a super-interface with these two methods and have the interceptors inherit from it.

The challenge is naming it 😄 

Suggestions?

Reply:
It was [my suggestion on the issue](https://github.com/spring-projects/spring-kafka/issues/1704#issuecomment-896997190).

However, I see that we could/should extract a super-interface with these two methods and have the interceptors inherit from it.

The challenge is naming it 😄 

`BeforeAfterPollProcessor` sounds ok to me.

Reply:
That will work for me, too.
So, the logic in the Listener Container would be much simpler.

Reply:
Works for me, I'll add `BeforeAfterPollProcessor`